Sources/SwiftLexicalLookup/SequentialScopeSyntax.swift

```diff
@@ -0,0 +1,93 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Scope that, in addition to names introduced by itself,
+/// also handles names introduced by
+/// `IntroducingToSequentialParentScopeSyntax` children scopes.
+protocol SequentialScopeSyntax: ScopeSyntax {
+  /// Returns names introduced by `codeBlockItems`
+  /// and included `IntroducingToSequentialParentScopeSyntax` children
+  /// scopes that match the lookup.
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
+    for identifier: Identifier?,
+    at syntax: SyntaxProtocol,
+    with config: LookupConfig,
+    state: LookupState,
+    createResultsForThisScopeWith getResults: ([LookupName]) -> (LookupResult)
+  ) -> [LookupResult]
+}
+
+extension SequentialScopeSyntax {
+  func sequentialLookup(
+    in codeBlockItems: any Collection<CodeBlockItemSyntax>,
```

Comment:
It's because of how we currently handle the file scope i.e. we treat the part before the first non-declaration as a member block scope and then we use sequential lookup for the other part. That's why we need to be able to pass here just a partition of all the available code block items.

Though it might need to get changed, as I mentioned in [this comment](https://github.com/swiftlang/swift-syntax/pull/2748#discussion_r1700084713), it seems like the visibility in playgrounds is a bit different than what we've modeled so far.